### DISCUSSION TITLE
formatting number output

### DISCUSSION THREAD

**Discussion by mophilly** [enoch3712/ExtractThinker#138] [Q&A] [answered]
Using version 0.0.29. 
A field in the pydantic class definition:
``` owner_interest: float = Field(
            description="The decimal interest of the owner in the transaction, column name: 'Owner Interest' position: 7")
```
The value returned is 1.254e-05 instead of the decimal representation of 0.00001254. 
What is the preferred way to format the return value? E.g. write a validator or ?

**Comment by enoch3712** [comment, marked as answer]
Hello @mophilly! 

This is the LLM being creative. You can force it by given examples:

`            description="The decimal interest of the owner in the transaction, column name: 'Owner Interest' position: 7. Always return normal decimal and NEVER Scientific Notation")`

you can also implement a validator for this, if you want an extra layer of security

```
class InvoiceLine(BaseModel):
    description: str
    quantity: int
    unit_price: float
    amount: float

    @field_validator('quantity', mode='before')
    def convert_quantity_to_int(cls, v):
        if isinstance(v, float):
            return int(v)
        return v
```

Change the validator at will

**Comment by mophilly** [comment]
Here is what working for me right now. 
I have the Pydantic classes defined in a single .py file. 

The import statements:
```
from typing import List, Optional
from pydantic import Field, condecimal
from extract_thinker.models.contract import Contract  # the same as pydantic.BaseModel
```

Then in the class definitions where I need financial precision, I changed "float" to "condecimal". Note the instruction regarding white space. I tried a couple of forms before this worked with gpt-4o.
```
owner_taxes: condecimal(max_digits=10, decimal_places=2) = Field(description="The taxes of the owner, column name: 'Owner Taxes' position: 2. Important: There may be empty space under this column; in that case, return 0.")
```

Now that the result is at a good point, I am going to run gpt-4o-mini to see how its results differ from gpt-4o.

@enoch3712 , thank you for all the help and for putting this project together. It is really helping me move my project forward.

**Reply by enoch3712** [reply]
Very nice! 

You are welcome :)
